objects/tag.c

```diff
@@ -220,16 +221,52 @@ is_client_tagged(client_t *c, tag_t *t)
     return false;
 }
 
-/** Get the index of the first selected tag.
- * \return Its index.
+/** Get the index of the first selected tag on focused desktop
+ * \return Its index
  */
 int
-tags_get_first_selected_index(void)
+tags_get_first_selected_index_on_focused_screen(void)
 {
-    foreach(tag, globalconf.tags)
+    /* 1) Get focused screen as screen that has cursor on it.
+     * 2) Find first selected tag on that screen.
+     *    (tag is on screen if it has clients with that screen value)
+     * 3) If nothing found, then just get first selected (fallback)
+     * 4) Return index of found tag
+     */
+     tag_t **first_tag = NULL;
+     tag_t **target_tag = NULL;
+     screen_t *screen = screen_get_focused();
+     foreach(tag, globalconf.tags)
+     {
         if((*tag)->selected)
-            return tag_array_indexof(&globalconf.tags, tag);
-    return 0;
+        {
+            if(!first_tag)
+                first_tag = tag;
+            if(is_tag_on_screen(tag, screen))
+            {
+                target_tag = tag;
+                break;
+            }
+        }
+     }
+     if(target_tag)
+        return tag_array_indexof(&globalconf.tags, target_tag);
+     else
+        return tag_array_indexof(&globalconf.tags, first_tag);
+}
+
+/** Check if tag has clients on specified screen.
+ * \param t the tag
+ * \param s the screen
+ * \return true if the are client is tagged with the tag, false otherwise.
+ */
+bool
+is_tag_on_screen(tag_t **t, screen_t *s)
+{
```

Comment:
Can you make this `static inline` and remove it from the header? It's only used internally.

(Even better would be to get rid of it completely, but I don't think that you'll like that idea...)